pyangbind/lib/serialise.py

```diff
@@ -164,54 +160,271 @@ def default(self, obj, mode="default"):
+                # NOTE: previously with IETF mode the code would fall-through if obj was falsey
+                return self.yangt_empty(obj)
             else:
-                if obj:
-                    return True
-                else:
-                    return False
+                return bool(obj)
         elif map_val in ["pyangbind.lib.yangtypes.TypedList"]:
-            return [self.default(i) for i in obj]
-        elif map_val in ["int"]:
-            int_size = getattr(obj, "_restricted_int_size", None)
-            if mode == "ietf" and int_size == 64:
-                return six.text_type(obj)
-            return int(obj)
-        elif map_val in ["long"]:
+            return self.yangt_typed_list(obj)
+        elif map_val in ["int", "long"]:
             int_size = getattr(obj, "_restricted_int_size", None)
-            if mode == "ietf" and int_size == 64:
-                return six.text_type(obj)
-            elif six.PY3:
-                return int(obj)
-            else:
-                return long(obj)
+            return self.yangt_long(obj) if int_size == 64 else self.yangt_int(obj)
         elif map_val in ["container"]:
-            return self._preprocess_element(obj.get(), mode=mode)
+            return self.preprocess_element(obj.get())
         elif map_val in ["decimal.Decimal"]:
-            return six.text_type(obj) if mode == "ietf" else float(obj)
+            return self.yangt_decimal(obj)
+
+    def yangt_int(self, obj):
+        # for values that are 32-bits and under..
+        return int(obj)
+
+    def yangt_long(self, obj):
+        # don't need to sepcial-case for PY3 because we assign `long = int` in the header
+        return long(obj)
+
+    def yangt_identityref(self, obj):
+        raise UnmappedItem
+
+    def yangt_decimal(self, obj):
+        # doesn't seem to need a distinction for non-64-bit decimal?
+        return float(obj)
+
+    def yangt_empty(self, obj):
+        return bool(obj)
+
+    def yangt_typed_list(self, obj):
+        return [self.default(i) for i in obj]
+
+
+class IETFYangDataSerialiser(YangDataSerialiser):
+    """
+    IETF data serialiser overrides some of the data type formats only
+    """
+
+    def yangt_long(self, obj):
+        return six.text_type(obj)
+
+    def yangt_identityref(self, obj):
+        try:
+            emod = obj._enumeration_dict[obj]["@module"]
+            if emod != obj._defining_module:
+                return "%s:%s" % (obj._enumeration_dict[obj]["@module"], obj)
+        except KeyError:
+            pass
+        return six.text_type(obj)
+
+    def yangt_decimal(self, obj):
+        return six.text_type(obj)
+
+    def yangt_empty(self, obj):
+        return [None] if obj else False
+
+
+class XmlYangDataSerialiser(IETFYangDataSerialiser):
+    """
+    XML can have an empty tag, and a TypedList must be treated specially, here we mark it with a deque
+    """
+
+    def yangt_typed_list(self, obj):
```

Comment:
It would be worth adding some more commentary as to why this we need to return this as a `deque` rather than simply just returning it as a `list` as in the other implementations.

Reply:
I've just added a note about the use of *deque*. To expand on that: this is really an implementation detail. All I wanted is to be able to identify a typed list in the next phase of encoding. (Because we serialise containers into a list already, I can't use a list again to represent the typed list).

So I was looking for a _different_ type that hasn't been used. One is to use a set which should be more natural (because, e.g. a leaf-list should behave like a set -- unique items) but the set implementation doesn't maintain ordering and I didn't want to write an `OrderedSet` just for this. A `deque`  behaves like a list and maintains ordering so I used it.

Perhaps, we could actually flip this around => make a container a set (since ordering doesn't need to be maintained AFAIK?) and then => typed list as list ?

Reply:
This seems a bit forced/contrived. I think this should use a custom declared type instead. For example, ```python
class CustomList(list):
    pass

return CustomList([self.default(i) for i in obj])
```
Of course, I haven't had a chance to look through this entire set of changes so I can't say for sure this is the right option. But using a new data type just to have something different doesn't feel right.

On a tangential note, this puts me in mind of the ORM-style enhancements I was working on recently. I wonder if, first, it wouldn't make sense to do a type revamp, and turn all of the internal types into actual real types with built-in coercion.

Reply:
Yep, that makes sense. I've replace the use of *deque* now with a list subclass.

The "ORM-style enhancements" sound interesting. Is that on a branch somewhere? But, I think XML serialisation is useful _now_ (at least for me, and probably many others using NETCONF) so I think we could do this first :)

Reply:
Cool, I think this is a good step forward.

The ORM stuff is in a very premature state right now, basically just theoretical/experimental. But I think that enhancements to the serialization could potentially help with it, if the types can all get properly abstracted out. And that can help lead to a cleaner library! Anyhow, if you want to get a look at those changes, they are here: https://github.com/robshakir/pyangbind/compare/master...tarkatronic:orm_refactor and if you want to talk further on them, feel free to join us on the NetworkToCode slack!

Reply:
Thanks for this change, I agree with Joey's assessment here.